### DISCUSSION TITLE
32h743 does not work under certain circumstances

### DISCUSSION THREAD

**Discussion by ghost** [hathach/tinyusb#1934] [Q&A]
Hi,
I am facing a strange behavior . I set TinyUSB to work with a CDC & MSC devices (like the proposed example in the package).
Now, I ported the example cdc_msc in my environment (RTOS on a Nucleo stm32743). Here I have a strange behavior . If tud_task(); is placed in a loop (using 100% of the cpu) the program works (I can communicate via the cdc and I can see the mounted disk on my desktop).
This works (but it uses 100% of the cpu)

```
        ...
        tud_init(BOARD_TUD_RHPORT);
	while (TRUE) {

// TinyUSB device task
		tud_task();
	}
        ...
```
This does not work ...
```
        ...
        tud_init(BOARD_TUD_RHPORT);
	while (TRUE) {

// Waiting for 500us (tud_task(); is called ~ every 500us)
                kern_suspedProcess(500);

// TinyUSB device task
		tud_task();
	}
        ...
```
Of course, the second example should be the good one;  kern_suspedProcess(500) suspend the process for 500us and change the context. With the scope I measured that tud_task(); is really called regularly @ 500us with a very little jitter .
Exactly the same code works perfectly on the Nucleo stm32l4r5.
Any ideas about this strange behavior ?
BR
Edo

**Comment by hathach** [comment]
which RTOS you are running ?

**Comment by ghost** [comment]
Hi, it is a professional non open source RTOS.
Btw, I have just did a simple test on your cdc_msc example (so no RTOS).
Your loop works well as long as I use 100% of the cpu ...
This is your main:
```
/*------------- MAIN -------------*/
int main(void)
{
  board_init();

  // init device stack on configured roothub port
  tud_init(BOARD_TUD_RHPORT);

  while (1)
  {
    tud_task(); // tinyusb device task
    led_blinking_task();

    cdc_task();
  }

  return 0;
}

```
Now, I have just added an WFI to be stopped until I receive an interruption (usb or 1ms SysTick). Now, something goes wrong with the tinyUSB.

```
/*------------- MAIN -------------*/
int main(void)
{
  board_init();

  // init device stack on configured roothub port
  tud_init(BOARD_TUD_RHPORT);

  while (1)
  {
__asm("wfi\n");

    tud_task(); // tinyusb device task
    led_blinking_task();

    cdc_task();
  }

  return 0;
}
 
```
Any idea? I tried other stm (not H7) and things work well
BR
Edo

**Reply by hathach** [reply]
have you added your rtos to the osal layer, and configure CFG_TUSB_OS accodingly ?

**Reply by ghost** [reply]
Yess, but the problem seems to be in another part of the code. See my last comment and test

**Reply by hathach** [reply]
You really need to compare your code against another RTOS such as freeRTOS. Using _wfi/wfe as your example is not safe, it will cause race condition between 'mcu about to go low power' and 'usb interrupt occur'. I have some detailed explanation on using wfi/wfe within this repo and/or [circuitpython](https://github.com/adafruit/circuitpython) that requires the use of  https://github.com/hathach/tinyusb/blob/master/src/osal/osal_none.h#L190 . You could try to find it if interested enough.

**Reply by ghost** [reply]
OK, thank you for the glue.
However I checked many comments you did in this repo. Specially the one #414  ..."Therefore we can call wfi, only when both TinyUSB task queue is empty and no more bottom half interrupt processing is left to do." ...
Well, I simply include **tud_task_event_ready** in your example with the hope that the WFI now works correctly. No way, same problem. So, to be honest I do not understand why because the queue is empty before to move in a sleep mode.
Here is your example code modified for the test.

```
int main(void) {
	board_init();

// init device stack on configured roothub port
	tud_init(BOARD_TUD_RHPORT);

	while (1) {
  
		__asm("cpsid i");
		if (tud_task_event_ready() == false) {

// The queue is empty, go to sleep and waiting for an interruption

			__asm("wfi");
		}
		__asm("cpsie i");
		tud_task(); // tinyusb device task
	
		led_blinking_task();

		cdc_task();
	}

	return 0;
}
```

So, it doesn't work.
Question, in your simple example how to manage correctly the WFI?

**Reply by ghost** [reply]
hathach, any suggestions about how to include a WFI in your simple code? Unfortunately,  H743 does not work properly (your simple example, with WFI and tud_task_event_ready; hangs!!). On other targets it works like a charm. What is wrong with TinyUSB & 32H743

**Reply by hathach** [reply]
If you are looking for exact code, try to find my PR to circuitpython regarding the low power feature for nrf52 or samd as mentioned above. Since I don't really remember everything.

**Reply by ghost** [reply]
hathach, thank you for your reply. The problem is not the generic low power features. The problem is the low power feature for the stm32H743. I tried many other architectures and all work fine. What I suspect is that there is a problem with the  stm32H743 implementation. This is why I came back to your original and very simple example ... that does not work.

**Reply by hathach** [reply]
I don't know why it works for other cortex and not with stm32h743. It can be a specific setting of the MCU. For that you should try to go through h7 manual/datasheet to troubleshoot the issue. My knowledge is pretty much limitted to usb and related work.

**Reply by BitLoose** [reply]
Did you solve the problem?  There are posts on the STM Community about the WFI sleep functions not working as per the manual, not sure if a latency issue might affect your code.
I'm also looking for a dual role USB driver for the STM32H743 as STM have only very crude drivers that won't even managed a mouse and keyboard at the same time, sigh

**Reply by ghost** [reply]
Hi BitLoose,
I found the problem time ago (I discovered the solution in a ST blog discussion).
In the usb initialisation it is mandatory to add this instruction:

RCC->AHB1LPENR &= ~RCC_AHB1LPENR_USB2OTGHSULPILPEN;

This was communicated to Hathach for updating the package.
Now the H743 (and other H7) works perfectly.
